rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,728 @@
+---
+authors: Tiago Silva (tiago.silva@goteleport.com)
+state: draft
+---
+
+# RFD XX - Kubernetes Per-pod RBAC
+
+## Required Approvers
+
+- Engineering: `@r0mant`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden`
+
+## What
+
+This RFD proposes a just-in-time access request to Kubernetes Pods that will
+allow users to request access to specific Pods and a static process for Teleport
+administrators to limit the access that a particular Role has to certain Pods.
+
+### Related issues
+
+- TBD
+
+## Why
+
+Teleport does not manage individual Kubernetes RBAC rules assigned to users or roles.
+Instead, Teleport allows users to impersonate a defined set of Kubernetes RBAC
+principals - groups or users - existing in the target cluster. Each user request
+will include the associated principals as impersonation headers, and Kubernetes
+RBAC will use them to control whether the request can be allowed.
+
+To change the user's permissions, you must create or edit a Kubernetes RBAC role
+in the Kubernetes cluster and associate it with an RBAC principal. Afterward, you
+append it into the Teleport user properties and the user can impersonate it.
+Currently, Teleport appends the RBAC principals to Teleport Roles or to the user
+itself. It means that if a user wants temporary access to a single Kubernetes POD,
+the cluster administrator has to create a Kubernetes RBAC Role, assign it to a
+Kubernetes RBAC principal and create a role with the principal. After that, the user
+has to request access to the temporary role. Once someone approves the role access
+request, the user can impersonate it.
+
+In addition to the complex process, limiting access to pods using Kubernetes RBAC
+has restrictions because Kubernetes RBAC does not support controlling access
+resources using pattern matching.
+The RBAC Role must include the resource name to grant access, but it requires
+knowing the Pod name in advance. In many cases, it is impossible to know them
+because Pods generated from Kubernetes objects like Deployments include random
+strings in their names and change each time the Pod is deleted or relocated from
+one node to the other.
+
+This RFD proposes an extension to Teleport RBAC where it will support
+managing access to Pods - including Pods names following a pattern - without
+relying on requesting access to Teleport Roles.
+
+## Role definition
+
+This RFD proposes an alternative to limit or extend user access to certain pods
+in a Kubernetes cluster from a Teleport Role.
+
+To achieve this, it proposes changes to Teleport's Roles to introduce a new
+Role version - `v6` - that extends the `v5` to include a field `kubernetes_pods`
+where it's possible to specify the Pod names to allow access.
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: my-kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_pods:
```

Comment:
What would be the value of `kubernetes_resources`? A list of resource URIs like `/apis/GROUP/VERSION/RESOURCETYPE/NAME`?

Reply:
I forgot to update the RFD's references to `kubernetes_pods` although the RoleV6 PR only has `kubernetes_resources` 

@ptgott  the format is the following:

```yaml
kubernetes_resources:
- kind: pod
   name: "pod1"
   namespace: "namespace1"
- kind: pod
   name: "pod-*"
   namespace: "custom_namespace"

We only support `kind:pod` but it can be extended in the future.